### ISSUE TITLE
CPU training stalls for 60s+ compiling native artifact for LightGBM predictor kernel

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#277]
Earliest public failure is recipe.train().run() remaining active after 60s on backend cpu. Composition uses recipe.model().lgbm().norm(batch).iq(2).xxs followed by conv(8, 1).tan().norm(batch).qi(4).1 with loss(ce), and recipe.data("data/categoric/sample_subfolders").target("target").split(0.8). The load is moderate (~1500 rows, feature count from folder structure) with a small LightGBM tree ensemble (default boost_iterations / lightgbm_leaves) and a single quantized conv block; arithmetic and memory traffic are far below 60s on a CPU. The stall occurs at lower_estimator -> fit_lightgbm (recipe.rs:7437) -> boosted_predictor (7344) -> compile_native_artifact (2828) -> native_cpu_compiler for the Predictor primitive. No existing open issue represents the LightGBM predictor compilation path (252 is CatBoost cbst, 255 is Forest/XGBoost, 254 is Bayes/perceptron, 253 MoE). Measured 60s is materially slower than the workload and hardware justify; expected is seconds.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `39b1acbc659733d9`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"new","issue":0,"title":"CPU training stalls for 60s+ compiling native artifact for LightGBM predictor kernel","body":"Earliest public failure is recipe.train().run() remaining active after 60s on backend cpu. Composition uses recipe.model().lgbm().norm(batch).iq(2).xxs followed by conv(8, 1).tan().norm(batch).qi(4).1 with loss(ce), and recipe.data(\"data/categoric/sample_subfolders\").target(\"target\").split(0.8). The load is moderate (~1500 rows, feature count from folder structure) with a small LightGBM tree ensemble (default boost_iterations / lightgbm_leaves) and a single quantized conv block; arithmetic and memory traffic are far below 60s on a CPU. The stall occurs at lower_estimator -> fit_lightgbm (recipe.rs:7437) -> boosted_predictor (7344) -> compile_native_artifact (2828) -> native_cpu_compiler for the Predictor primitive. No existing open issue represents the LightGBM predictor compilation path (252 is CatBoost cbst, 255 is Forest/XGBoost, 254 is Bayes/perceptron, 253 MoE). Measured 60s is materially slower than the workload and hardware justify; expected is seconds.","rationale":"Inspected issues 252, 255, 254, 253, 250, 164, 167, 210 and source paths fit_lightgbm/boosted_predictor/compile_native_artifact. Packet reaches the same earliest failing operation (recipe.train().run stall during native CPU artifact compilation) but through fit_lightgbm, which is not covered by any existing issue. New input representation (categoric subfolder data with lgbm().norm(batch).iq(2).xxs) is ignored by the same loader and produces the same terminal check, but the defect is the unreported LightGBM predictor compilation stall, so it establishes a distinct earliest cause."}
```

## Deterministic failure packet

kind=performance
id=39b1acbc659733d9
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:11188 next:11189 composition:120164728256732
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-232-11188/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/categoric/sample_subfolders")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.lgbm().norm(batch).iq(2).xxs
		.conv(8, 1).tan().norm(batch).qi(4).1
		.loss(ce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(120164728256732)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Failed on Saturday, August 29, 2026 11:46 PM

Commit: 9098de8711552004bb9eabd84c6d8d9c120d4c58

Command:
/usr/bin/time -f 'elapsed_seconds=%e max_rss_kb=%M' timeout 65s env PATH=/home/nate/Desktop/recipe-minimal/target/release:$PATH RECIPE_FORCE_CPU=1 recipe test.rs

Reproduction script:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("/home/nate/Desktop/recipe-minimal/data/categoric/sample_subfolders")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.lgbm().norm(batch).iq(2).xxs
		.conv(8, 1).tan().norm(batch).qi(4).1
		.loss(ce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(120164728256732)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```

Full output:
```text
thread 'main' (352212) panicked at recipe.rs:12863:13:
dataset has no training features
note: run with RUST_BACKTRACE=1 environment variable to display a backtrace
Command exited with non-zero status 101
elapsed_seconds=0.11 max_rss_kb=148228
```

Verdict: FAIL
Reason: The current public path does not reach LightGBM compilation or complete training. The canonical folder dataset is rejected because it has no training features, so the reported 60-second compiler stall and the expected finite one-epoch result remain unproven.

**Comment by nm-z** [comment]
Earliest public failure is `recipe.train().run()` remaining active past the slow-runtime threshold on `backend:cpu`. The measured 107 seconds is materially slower than the single-digit seconds justified by the workload. The dominant gap is `compile_native_artifact` covering the LightGBM predictor kernel. The trailing `.layer(8)` and `.moe(1, [layer(8), layer(8)]).selu().norm(batch).qi(4).k.s` blocks reach the same compilation invocation, expanding issue 277 with additional evidence.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `242d14da7b09449d`
- Classifier provider: `agy`
- Classifier model: `gemini-3.1-pro-high`
- Classifier effort: `high`

### Classifier decision

```json
{"body":"Earliest public failure is `recipe.train().run()` remaining active past the slow-runtime threshold on `backend:cpu`. The measured 107 seconds is materially slower than the single-digit seconds justified by the workload. The dominant gap is `compile_native_artifact` covering the LightGBM predictor kernel. The trailing `.layer(8)` and `.moe(1, [layer(8), layer(8)]).selu().norm(batch).qi(4).k.s` blocks reach the same compilation invocation, expanding issue 277 with additional evidence.","issue":277,"rationale":"Derived workload from `data/numeric/semicolon_data.data`: 550,230 rows and 1 feature. The `recipe.model()` starts with `.lgbm().cos().qi(5).k.m`. With default 8 boost iterations and 8 leaves, tree candidates and depth are small, requiring roughly 70M scalar operations for `fit_lightgbm`. Arithmetic work for the appended `.layer(8)` and `.moe(1, [layer(8), layer(8)]).selu().norm(batch).qi(4).k.s` is minimal (sub 0.1 GFLOP) on CPU. Total memory traffic per epoch is ~4.4 MB. Given modern CPU capabilities, a defensible expected runtime is single-digit seconds. The measured 107 seconds is materially slower. The dominant gap is train preparation, specifically the `compile_native_artifact` clang stall for the LightGBM predictor kernel. Read issues 277, 253, and 357. The LightGBM compilation stall (277) is reached first and dominates the preparation time, so this is repeated evidence for issue 277.","title":"CPU training stalls for 60s+ compiling native artifact for LightGBM predictor kernel","verdict":"comment"}
```

## Deterministic failure packet

kind=performance
id=242d14da7b09449d
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:16000 next:16001 composition:97377517961600
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 107 seconds
measurement=elapsed_seconds:107 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-4-16000/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/numeric/semicolon_data.data")
		.target("target");
	let model = recipe.model()
		.lgbm().cos().qi(5).k.m
		.layer(8)
		.moe(1, [layer(8), layer(8)]).selu().norm(batch).qi(4).k.s
		.loss(rmse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(97377517961600)
		.epochs(1)
		.log(all)
		.int(8)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Re-verified on `minimal` b384815c, archy cpu, machine load 25-28, cold native cache per run: `RECIPE_FORCE_CPU=1 recipe reproduction.rs` with `lgbm(8).norm(batch).iq(2).xxs.conv(8, 1).tan().norm(batch).qi(4).1` on data/categoric/sample_subfolders split 0.8, one epoch: 5.3 s and 5.7 s in two runs, loss 0.727209909021268 both, two native artifacts. Margin against the 60 s threshold is 10x; the LightGBM predictor kernel compile the packet blamed is gone since #556 and #541. No residual defect on the path, so no new PR.

**Comment by nm-z** [comment]
Closing: no longer reproducible under the measured conditions. Re-verified 2026-09-12 on `minimal` b384815c (archy cpu, load 25-28, cold native cache per run): the packet's reproduction completes one epoch in 5.3-5.7 s, 10x under the 60 s threshold; the LightGBM predictor kernel compile it blamed is gone since #556 (40b92c0c) and #541 (a6247877), both on current `minimal` 3e278905. This is not a claim that every CPU compile is fast; reopen with a packet that crosses 60 s on current `minimal`.
